File: apps/tauri/src-tauri/src/types/history.rs
//! Query history types
//!
//! Types for query history storage and retrieval.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryHistoryEntry {
    pub id: String,
    pub db_path: String,
    pub query_text: Option<String>,
    pub query: Option<String>,
    pub executed_at: Option<String>,
    pub timestamp: Option<String>,
    pub duration_ms: Option<u64>,
    pub success: Option<bool>,
    pub error: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetQueryHistoryRequest {
    pub db_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetQueryHistoryResponse {
    pub success: bool,
    pub history: Option<Vec<QueryHistoryEntry>>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveQueryHistoryRequest {
    pub entry: Option<QueryHistoryEntry>,
    pub query: Option<String>,
    pub connection_path: Option<String>,
    pub timestamp: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveQueryHistoryResponse {
    pub success: bool,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteQueryHistoryRequest {
    pub db_path: Option<String>,
    pub entry_id: Option<String>,
    pub id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteQueryHistoryResponse {
    pub success: bool,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClearQueryHistoryRequest {
    pub db_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClearQueryHistoryResponse {
    pub success: bool,
    pub error: Option<String>,
}

File: apps/electron/src/main/services/ipc/history.ts
import type {
  ClearQueryHistoryRequest,
  DeleteQueryHistoryRequest,
  GetQueryHistoryRequest,
  QueryHistoryEntry,
  SaveQueryHistoryRequest,
} from '@shared/types';
import { IPC_CHANNELS } from '@shared/types';
import { ipcMain } from 'electron';
import {
  clearQueryHistory,
  deleteQueryHistoryEntry,
  getQueryHistory,
  saveQueryHistoryEntry,
} from '../store';
import { createHandler } from './utils';

export function setupHistoryHandlers(): void {
  // Query History: Get
  ipcMain.handle(
    IPC_CHANNELS.QUERY_HISTORY_GET,
    createHandler(async (request: GetQueryHistoryRequest) => {
      const dbPath = request.dbPath || '';
      const history = getQueryHistory(dbPath);
      return { success: true, history };
    })
  );

  // Query History: Save
  ipcMain.handle(
    IPC_CHANNELS.QUERY_HISTORY_SAVE,
    createHandler(async (request: SaveQueryHistoryRequest) => {
      const entry: QueryHistoryEntry = request.entry || {
        id: crypto.randomUUID(),
        query: request.query || '',
        dbPath: request.connectionPath || '',
        timestamp: request.timestamp || new Date().toISOString(),
        description: request.description,
      };
      saveQueryHistoryEntry(entry);
      return { success: true, entry };
    })
  );

  // Query History: Delete
  ipcMain.handle(
    IPC_CHANNELS.QUERY_HISTORY_DELETE,
    createHandler(async (request: DeleteQueryHistoryRequest) => {
      const dbPath = request.dbPath || '';
      const entryId = request.id || request.entryId || '';
      deleteQueryHistoryEntry(dbPath, entryId);
      return { success: true };
    })
  );

  // Query History: Clear
  ipcMain.handle(
    IPC_CHANNELS.QUERY_HISTORY_CLEAR,
    createHandler(async (request: ClearQueryHistoryRequest) => {
      const dbPath = request.dbPath || '';
      clearQueryHistory(dbPath);
      return { success: true };
    })
  );
}
